src/Microsoft.ML.Core/Data/ReadOnlyMemoryUtils.cs

```diff
@@ -0,0 +1,397 @@
+                for (int ichCur = 0; ;)
+                {
+                    int ichMinLocal = ichCur;
+                    for (; ; ichCur++)
+                    {
+                        Contracts.Assert(ichCur <= memory.Length);
+                        if (ichCur >= memory.Length)
+                        {
+                            yield return memory.Slice(ichMinLocal, ichCur - ichMinLocal);
+                            yield break;
+                        }
+                        if (memory.Span[ichCur] == chSep)
+                            break;
+                    }
+
+                    yield return memory.Slice(ichMinLocal, ichCur - ichMinLocal);
+
+                    // Skip the separator.
+                    ichCur++;
+                }
+            }
+            else
+            {
+                for (int ichCur = 0; ;)
+                {
+                    int ichMinLocal = ichCur;
+                    for (; ; ichCur++)
+                    {
+                        Contracts.Assert(ichCur <= memory.Length);
+                        if (ichCur >= memory.Length)
+                        {
+                            yield return memory.Slice(ichMinLocal, ichCur - ichMinLocal);
+                            yield break;
+                        }
+                        // REVIEW: Can this be faster?
+                        if (ContainsChar(memory.Span[ichCur], separators))
+                            break;
+                    }
+
+                    yield return memory.Slice(ichMinLocal, ichCur - ichMinLocal);
+
+                    // Skip the separator.
+                    ichCur++;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Splits <paramref name="memory"/> on the left-most occurrence of separator and produces the left
+        /// and right <see cref="ReadOnlyMemory{T}"/> of <see cref="char"/> values. If <paramref name="memory"/> does not contain the separator character,
+        /// this returns false and sets <paramref name="left"/> to this instance and <paramref name="right"/>
+        /// to the default <see cref="ReadOnlyMemory{T}"/> of <see cref="char"/> value.
+        /// </summary>
+        public static bool SplitOne(char separator, out ReadOnlyMemory<char> left, out ReadOnlyMemory<char> right, ReadOnlyMemory<char> memory)
+        {
+            if (memory.IsEmpty)
+            {
+                left = memory;
+                right = default;
+                return false;
+            }
+
+            var span = memory.Span;
+            int ichCur = 0;
+            for (; ; ichCur++)
+            {
+                Contracts.Assert(0 <= ichCur && ichCur <= memory.Length);
+                if (ichCur >= memory.Length)
+                {
+                    left = memory;
+                    right = default;
+                    return false;
+                }
+                if (span[ichCur] == separator)
```

Comment:
Should use `MemoryExtensions.IndexOf` instead. It is hardware accelerated.

Reply:
These take Span as input and in our case we have character(s) as input so I will have to convert character to span which I think is expensive because AsMemory() seemed expensive so may be AsSpan might also be?

---
In reply to: [216468406](https://github.com/dotnet/machinelearning/pull/863#discussion_r216468406) [](ancestors = 216468406)

Reply:
There is an IndexOf overload that takes a single character.  

```
public static int IndexOf<T>(this System.Span<T> span, T value) where T : System.IEquatable<T>
```

---
In reply to: [216845370](https://github.com/dotnet/machinelearning/pull/863#discussion_r216845370) [](ancestors = 216845370,216468406)